File: lab4_ADS/problemc.cpp
#include <iostream>
using namespace std;
struct Node {
    int data;
    Node* left;
    Node* right;
    Node(int val) : data(val), left(nullptr), right(nullptr) {}
};

Node* insert(Node* root, int val) {
    if (!root) return new Node(val);
    if (val < root->data)
        root->left = insert(root->left, val);
    else
        root->right = insert(root->right, val);
    return root;
}


Node* find(Node* root, int k) {
    if (!root || root->data == k)
        return root;
    if (k < root->data)
        return find(root->left, k);
    else
        return find(root->right, k);
}

void preorder(Node* root) {
    if (!root) return;
    cout << root->data << " ";
    preorder(root->left);
    preorder(root->right);
}

int main() {
    int n;
    cin >> n;
    int gifts[n];

    for (int i = 0; i < n; i++) {
        cin >> gifts[i];
    }

    int k;
    cin >> k;

    Node* root = nullptr;
    for (int i = 0; i < n; i++) {
        root = insert(root, gifts[i]);
    }

    Node* k_node = find(root, k);

    preorder(k_node);
    cout << endl;

    return 0;
}

File: lab4_ADS/problemc.py
class TreeNode:
    def __init__(self, value):
        self.value = value
        self.left = None
        self.right = None

def insert_into_bst(root, value):
    if root is None:
        return TreeNode(value)
    if value < root.value:
        root.left = insert_into_bst(root.left, value)
    else:
        root.right = insert_into_bst(root.right, value)
    return root

def find_subtree(root, k):
    if root is None or root.value == k:
        return root
    if k < root.value:
        return find_subtree(root.left, k)
    else:
        return find_subtree(root.right, k)

def preorder_traversal(root, result):
    if root:
        result.append(root.value)
        preorder_traversal(root.left, result)
        preorder_traversal(root.right, result)


n = int(input())
gifts = list(map(int, input().split()))
k = int(input())


root = None
for gift in gifts:
    root = insert_into_bst(root, gift)


subtree_root = find_subtree(root, k)


result = []
preorder_traversal(subtree_root, result)


print(" ".join(map(str, result)))
